File: 318-maximum-product-of-word-lengths/318-maximum-product-of-word-lengths.py
class Solution:
    def helper(self, s1, s2):
        for ch in s1:
            if ch in s2:
                return False
        return True
        
    def maxProduct(self, words: List[str]) -> int:
        n = len(words)
        
        ans = 0
        s = [set() for _ in range(n)]
        
        for i in range(n):
            for ch in words[i]:
                s[i].add(ch)
            
            for j in range(i):
                if self.helper(s[i], s[j]):
                    ans = max(ans, len(words[i]) * len(words[j]))
        return ans

File: 318-maximum-product-of-word-lengths/318-maximum-product-of-word-lengths.cpp
class Solution {
public:
    int maxProduct(vector<string>& words) {
        return solve(words);
    }
private:
    // daily
    bool helper(unordered_set<char>& s1, unordered_set<char>& s2) {
        for(auto ch: s1) {
            if(s2.find(ch) != s2.end())
                return false;
        }
        return true;
    }
    int solve(vector<string>& words) {
        int n = words.size();
        int ans = 0;
        vector<unordered_set<char>> s(n);
        
        for(int i = 0; i < n; ++i) {
            
            for(char ch: words[i])
                s[i].insert(ch);
            for(int j = 0; j < i; ++j) {
                if(helper(s[i], s[j]))
                    ans = max(ans, (int)(words[i].length() * words[j].length()));
            }
        } 
        return ans;
    }
};